Add Confidential tab

ericglau opened the pull request:
Adds a Confidential tab that implements a confidential fungible token, using OpenZeppelin Confidential Contracts.

Includes:  
- `@openzeppelin/wizard-confidential` package, UI, AI assistant, MCP tools.

ericglau commented:
@SocketSecurity ignore npm/prettier@3.6.2

ericglau commented on packages/core/confidential/src/zip-hardhat.test.ts.md:
```diff
@@ -0,0 +1,234 @@
+          "@fhevm/solidity": "0.7.0",␊
+          "@zama-fhe/relayer-sdk": "^0.1.0-5",␊
+          "@nomicfoundation/hardhat-toolbox": "^6.1.0",␊
+          "hardhat": "^2.16.1"␊
+        }␊
```
this is just the minimum version, the lockfile is used for pinning

ericglau commented on packages/mcp/src/confidential/schemas.ts:
```diff
@@ -0,0 +1,46 @@
+    .or(z.literal('zama-ethereum'))
+    .describe(confidentialConfidentialFungibleDescriptions.networkConfig),
+  wrappable: z.boolean().optional().describe(confidentialConfidentialFungibleDescriptions.wrappable),
+  votes: z
+    .literal('blocknumber')
+    .or(z.literal('timestamp'))
+    .optional()
+    .describe(confidentialConfidentialFungibleDescriptions.votes),
+  ...commonSchema,
+} as const satisfies z.ZodRawShape;
```
This schema is for the MCP tool definition, and `undefined` is functionally equivalent to `false`, so it is sufficient.

ericglau commented on packages/ui/api/ai-assistant/function-definitions/confidential.ts:
```diff
@@ -0,0 +1,43 @@
+        type: 'boolean',
+        description: confidentialConfidentialFungibleDescriptions.wrappable,
+      },
+      votes: {
+        anyOf: [
+          { type: 'boolean', enum: [false, true] },
+          { type: 'string', enum: ['blocknumber', 'timestamp'] },
+        ],
+        description: confidentialConfidentialFungibleDescriptions.votes,
+      },
```
The AI Assistant in Wizard does not use the MCP tools, but defines the function definitions for itself.

ericglau commented on packages/ui/src/main.ts:
```diff
@@ -5,6 +5,7 @@ import CairoApp from './cairo/App.svelte';
 import CairoAlphaApp from './cairo_alpha/App.svelte';
 import StellarApp from './stellar/App.svelte';
 import StylusApp from './stylus/App.svelte';
+import ConfidentialApp from './confidential/App.svelte';
```
The imports come from local paths that are part of the monorepo, so it does not strictly need these dependencies in package.json

ericglau commented on packages/core/confidential/src/zip-hardhat.test.ts.md:
```diff
@@ -0,0 +1,234 @@
+              ConfidentialFungibleTokenERC20Wrapper(underlying)␊
+              EIP712("My Token", "1")␊
+          {␊
+              _mint(recipient, FHE.asEuint64(SafeCast.toUint64(2000 * 10 ** decimals())));␊
+              _grantRole(DEFAULT_ADMIN_ROLE, defaultAdmin);␊
+              _grantRole(HANDLE_VIEWER_ROLE, handleViewer);␊
+          }␊
+      ␊
+          function clock() public view override returns (uint48) {␊
```
For the Confidential Contracts library, `decimals()` is limited to 6

CoveMB commented:
<img width="311" height="478" alt="Screenshot 2025-09-24 at 5 21 25 PM" src="https://github.com/user-attachments/assets/88439a45-c0ec-45f0-98bc-4e77a2199beb" />

Not sure if we can make AI description more clear as it already mention confidential + erc-20 

Also wondering in tab using "confidential" vs "Zama" ?

CoveMB commented on packages/core/confidential/src/test.ts:
```diff
@@ -0,0 +1,34 @@
+    settings.outputSelection = {};
+  }
+
+  await hre.run('compile');
```
would a try catch .. finally block with cleanup be useful?

CoveMB reviewed (approved):
Nice work!

I tried the security review from MCP unittest which resulted in the following (are any relevant to the case?)

# Recommendations

  New confidential tooling is structurally sound, but one concurrency bug can leak per-contract metadata, and newly introduced pre-release
  FHE dependencies deserve extra due diligence.

  ## Risk Level Low

  UnvalidatedNetworkConfig
  location: packages/core/confidential/src/confidentialFungible.ts:45
  type: input validation / configuration hardening
  description: addNetworkConfig silently does nothing if networkConfig falls outside the hard-coded list. Typed callers are protected,
  but any untyped consumer (custom JS integration, AI tooling bugs) can emit a contract missing the required FHE network parent, leaving
  deployments misconfigured and functions reverting at runtime.
  recommendation: Throw an OptionsError when the value is not one of the supported configs so invalid input fails fast.

  PreReleaseFHEDependencies
  location: packages/core/confidential/package.json:27, packages/core/confidential/src/environments/hardhat/package.json:13
  type: dependency risk
  description: The new confidential package and Hardhat template depend on pre-release FHE artifacts (@fhevm/hardhat-plugin@0.0.1-6, @fhevm/
  core-contracts@0.7.0-12, @zama-fhe/relayer-sdk@0.1.0-5). These builds have limited hardening history, so upstream changes or latent
  vulnerabilities pose elevated supply-chain risk.
  recommendation: Perform targeted security review of these transitive dependencies, pin to vetted versions, and monitor for advisories
  before promoting the wizard to production use.

  --

 Suggested next steps: 
 1) add runtime validation for networkConfig; 
 2) schedule a dependency risk assessment (or version pinning) for the new FHE toolchain.
 
 
 //--
 
 # Tests Recommendations

  ## Risk Level High

  printContractVersioned
  location: packages/core/confidential/src/print-versioned.ts
  change type: added
  reason: New Remix printer rewrites @openzeppelin/* and @fhevm/* import paths using pinned versions; a regression would generate broken
  Remix projects (external tool integration).
  suggested tests:

  - Build a ConfidentialFungible contract with wrappable and votes enabled, run printContractVersioned, and assert the output replaces
    @openzeppelin/contracts, @openzeppelin/confidential-contracts, and @fhevm/solidity with the pinned versions from contract-version-pins.
  - Feed a contract that only imports third-party modules outside those prefixes and assert the printer leaves those paths untouched to
    prevent over-eager rewrites.

  ## Risk Level Medium

  injectHyperlinks
  location: packages/ui/src/confidential/inject-hyperlinks.ts
  change type: added
  reason: Regex-based HTML injection now covers confidential and FHEVM imports; mistakes would surface as broken or unsafe links in the code
  viewer (HTML sanitization risk).
  suggested tests:

  - Unit test the helper with sample highlighted code containing each supported import (@openzeppelin/contracts, @openzeppelin/confidential-
    contracts, @fhevm/solidity) and assert the generated URLs include the correct pinned versions.
  - Add a negative case using a path with ../ segments to confirm the guard still prevents hyperlink injection for disallowed patterns.

  evaluateSelection
  location: packages/ui/src/main.ts:66
  change type: modified
  reason: Added the confidential branch that gates whether the new wizard loads or shows the incompatibility banner; misrouting would block
  the feature.
  suggested tests:

  - Unit test evaluateSelection('confidential', undefined) and evaluateSelection('confidential', incompatibleVersion) to verify it
    respectively returns the ConfidentialApp path or an incompatible response with the expected semver hint.
  - Integration-style test stubbing new ConfidentialApp to ensure the dispatcher is invoked when the selection reports compatible: true for
    the confidential language.

ericglau commented on packages/core/confidential/src/utils/map-values.ts:
```diff
@@ -0,0 +1,8 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export function mapValues<K extends keyof any, V, W>(obj: Record<K, V>, fn: (val: V) => W): Record<K, W> {
```
Removed since this isn't used in this package.  (There is the same function in solidity package, but we haven't needed to maintain it so I don't see a need to change its implementation)

ericglau commented on packages/core/confidential/src/confidentialFungible.ts:
```diff
@@ -0,0 +1,244 @@
+   * Whether to keep track of historical balances for voting in on-chain governance, and optionally specify the clock mode.
+   * Setting `true` is equivalent to 'blocknumber'. Setting a clock mode implies voting is enabled.
+   */
+  votes?: boolean | ClockMode;
```
Good catch. We needed to support boolean in the Solidity models for backwards compatibility, but we can remove `true` here since it's a new API.

ericglau commented:
> security review from MCP unittest

- UnvalidatedNetworkConfig: fixed
- PreReleaseFHEDependencies: this is the current state of the supported upstream dependencies. Will be improved when they are supported in newer versions of Confidential Contracts.
- printContractVersioned: fixed
- injectHyperlinks, evaluateSelection: unnecessary, since any issues should be quite apparent in the UI

ericglau commented:
Notes on recent updates:
- Support Confidential Contracts v0.3.0: Renamed all API and UIs to ERC7984, changed tokenURI to contractURI
- Confidential Contracts v0.3.0 has a peer dependency on `@fhevm/solidity@0.9.1`, but the NPM package for `@fhevm/solidity@0.9.1` appears to correspond to GitHb tag [v0.9.14](https://github.com/zama-ai/fhevm/blob/v0.9.14/library-solidity/config/ZamaConfig.sol).  So we’ll use that NPM package, but will map import links in the UI to point users to v0.9.14.
- `SepoliaConfig` and `EthereumConfig` have been combined into `ZamaEthereumConfig`.  So the UI will only show 1 option under Network Configuration as “Zama Ethereum”, which is applicable to Ethereum mainnet and Sepolia.
